utils/utils.py

```diff
@@ -5,6 +5,23 @@
 import time
 
 
+if False:
```

Comment:
I needed this block of code to understand what happens behind the scenes, that's why I left it here even disabled. Would it make sense to add it with a debug flag?